File: android/student/src/main/java/it/scuola/registro/student/network/StudentApiService.kt
package it.scuola.registro.student.network

import it.scuola.registro.student.data.AttendanceRecord
import it.scuola.registro.student.data.GradeEntry
import it.scuola.registro.student.data.HomeworkAssignment
import it.scuola.registro.student.data.ScrutinyReportCard
import it.scuola.registro.student.data.StudentUser
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL

interface StudentApiService {
    suspend fun login(email: String, password: String): Result<Pair<String, StudentUser>>
    suspend fun getGrades(token: String): Result<List<GradeEntry>>
    suspend fun getAttendance(token: String): Result<List<AttendanceRecord>>
    suspend fun getHomework(token: String): Result<List<HomeworkAssignment>>
    suspend fun getReportCard(token: String): Result<ScrutinyReportCard>
}

class HttpStudentApiService(
    private val baseUrl: String = "https://registro-backend-fdu2.onrender.com/api/v1"
) : StudentApiService {

    override suspend fun login(email: String, password: String): Result<Pair<String, StudentUser>> =
        withContext(Dispatchers.IO) {
            try {
                var cleanEmail = email.trim().lowercase()
                if (cleanEmail.startsWith("studentea_")) {
                    cleanEmail = cleanEmail.replace("studentea_", "studente2a_")
                } else if (cleanEmail.startsWith("studenteb_")) {
                    cleanEmail = cleanEmail.replace("studenteb_", "studente2b_")
                }

                val url = URL("$baseUrl/auth/login")
                val conn = (url.openConnection() as HttpURLConnection).apply {
                    requestMethod = "POST"
                    setRequestProperty("Content-Type", "application/json; charset=UTF-8")
                    doOutput = true
                    doInput = true
                    connectTimeout = 10000
                    readTimeout = 10000
                }

                val payload = JSONObject().apply {
                    put("email", cleanEmail)
                    put("password", password)
                }

                OutputStreamWriter(conn.outputStream).use { it.write(payload.toString()) }

                val responseCode = conn.responseCode
                if (responseCode in 200..299) {
                    val response = BufferedReader(InputStreamReader(conn.inputStream)).use { it.readText() }
                    val json = JSONObject(response)
                    val userObj = json.optJSONObject("user") ?: JSONObject()
                    val role = userObj.optString("role")
                    if (role != "student") {
                        Result.failure(Exception("Accesso non consentito: questo account non appartiene a uno studente."))
                    } else {
                        val token = json.optString("token", json.optString("access_token"))
                        val user = StudentUser(
                            id = userObj.optString("id", "s1"),
                            firstName = userObj.optString("first_name", userObj.optString("name", "Studente")),
                            lastName = userObj.optString("last_name", ""),
                            email = userObj.optString("email", email),
                            className = userObj.optString("class_name", "Classe")
                        )
                        Result.success(Pair(token, user))
                    }
                } else {
                    val err = BufferedReader(InputStreamReader(conn.errorStream ?: conn.inputStream)).use { it.readText() }
                    Result.failure(Exception("Login fallito (HTTP $responseCode): $err"))
                }
            } catch (e: Exception) {
                Result.failure(e)
            }
        }

    override suspend fun getGrades(token: String): Result<List<GradeEntry>> =
        withContext(Dispatchers.IO) {
            try {
                val url = URL("$baseUrl/grades/my-grades")
                val conn = (url.openConnection() as HttpURLConnection).apply {
                    requestMethod = "GET"
                    setRequestProperty("Authorization", "Bearer $token")
                    setRequestProperty("Accept", "application/json")
                    connectTimeout = 10000
                    readTimeout = 10000
                }

                if (conn.responseCode in 200..299) {
                    val response = BufferedReader(InputStreamReader(conn.inputStream)).use { it.readText() }
                    val list = mutableListOf<GradeEntry>()
                    val trimmed = response.trim()
                    if (trimmed.startsWith("[")) {
                        val arr = JSONArray(trimmed)
                        for (i in 0 until arr.length()) {
                            val obj = arr.getJSONObject(i)
                            list.add(parseGradeEntry(obj, i))
                        }
                    } else if (trimmed.startsWith("{")) {
                        val root = JSONObject(trimmed)
                        val semesters = root.optJSONArray("semesters")
                        if (semesters != null) {
                            for (s in 0 until semesters.length()) {
                                val sObj = semesters.getJSONObject(s)
                                val gArr = sObj.optJSONArray("grades")
                                if (gArr != null) {
                                    for (g in 0 until gArr.length()) {
                                        list.add(parseGradeEntry(gArr.getJSONObject(g), list.size))
                                    }
                                }
                            }
                        } else {
                            val gArr = root.optJSONArray("grades")
                            if (gArr != null) {
                                for (g in 0 until gArr.length()) {
                                    list.add(parseGradeEntry(gArr.getJSONObject(g), list.size))
                                }
                            }
                        }
                    }
                    Result.success(list)
                } else {
                    Result.failure(Exception("Errore recupero voti: HTTP ${conn.responseCode}"))
                }
            } catch (e: Exception) {
                Result.failure(e)
            }
        }

    private fun parseGradeEntry(obj: JSONObject, index: Int): GradeEntry {
        val gradeVal = if (obj.has("grade_value")) obj.optDouble("grade_value", 0.0) else obj.optDouble("grade", 0.0)
        val subjectStr = obj.optString("subject_name", obj.optString("subject", "Materia"))
        val typeStr = obj.optString("evaluation_type", obj.optString("grade_type", obj.optString("type", "Orale")))
        var dateStr = obj.optString("date", "")
        if (dateStr.length >= 10) {
            dateStr = dateStr.substring(0, 10)
        }
        return GradeEntry(
            id = obj.optString("id", index.toString()),
            subject = subjectStr,
            grade = gradeVal,
            weight = obj.optDouble("weight", 1.0),
            type = typeStr,
            date = dateStr,
            period = obj.optInt("semester", obj.optInt("period", 1))
        )
    }

    override suspend fun getAttendance(token: String): Result<List<AttendanceRecord>> =
        withContext(Dispatchers.IO) {
            try {
                val url = URL("$baseUrl/attendance/my-attendance")
                val conn = (url.openConnection() as HttpURLConnection).apply {
                    requestMethod = "GET"
                    setRequestProperty("Authorization", "Bearer $token")
                    setRequestProperty("Accept", "application/json")
                    connectTimeout = 10000
                    readTimeout = 10000
                }

                if (conn.responseCode in 200..299) {
                    val response = BufferedReader(InputStreamReader(conn.inputStream)).use { it.readText() }.trim()
                    if (response == "null" || response.isEmpty()) {
                        return@withContext Result.success(emptyList())
                    }
                    val list = mutableListOf<AttendanceRecord>()
                    val arr = JSONArray(response)
                    for (i in 0 until arr.length()) {
                        val obj = arr.getJSONObject(i)
                        var dateStr = obj.optString("date", "")
                        if (dateStr.length >= 10) dateStr = dateStr.substring(0, 10)
                        list.add(
                            AttendanceRecord(
                                id = obj.optString("id", i.toString()),
                                date = dateStr,
                                type = obj.optString("status", obj.optString("type", "Presenza")),
                                isJustified = obj.optBoolean("is_justified", false),
                                reason = obj.optString("justification_reason", obj.optString("reason", ""))
                            )
                        )
                    }
                    Result.success(list)
                } else {
                    Result.failure(Exception("Errore recupero presenze: HTTP ${conn.responseCode}"))
                }
            } catch (e: Exception) {
                Result.failure(e)
            }
        }

    override suspend fun getHomework(token: String): Result<List<HomeworkAssignment>> =
        withContext(Dispatchers.IO) {
            try {
                val url = URL("$baseUrl/agenda")
                val conn = (url.openConnection() as HttpURLConnection).apply {
                    requestMethod = "GET"
                    setRequestProperty("Authorization", "Bearer $token")
                    setRequestProperty("Accept", "application/json")
                    connectTimeout = 10000
                    readTimeout = 10000
                }

                if (conn.responseCode in 200..299) {
                    val response = BufferedReader(InputStreamReader(conn.inputStream)).use { it.readText() }.trim()
                    if (response == "null" || response.isEmpty()) {
                        return@withContext Result.success(emptyList())
                    }
                    val list = mutableListOf<HomeworkAssignment>()
                    val arr = JSONArray(response)
                    for (i in 0 until arr.length()) {
                        val obj = arr.getJSONObject(i)
                        var dueStr = obj.optString("due_date", obj.optString("date", ""))
                        if (dueStr.length >= 10) dueStr = dueStr.substring(0, 10)
                        list.add(
                            HomeworkAssignment(
                                id = obj.optString("id", i.toString()),
                                subject = obj.optString("subject_name", obj.optString("subject", obj.optString("title", "Compito"))),
                                title = obj.optString("title", ""),
                                description = obj.optString("description", obj.optString("notes", "")),
                                dueDate = dueStr,
                                isCompleted = obj.optBoolean("is_completed", obj.optBoolean("completed", false))
                            )
                        )
                    }
                    Result.success(list)
                } else {
                    Result.failure(Exception("Errore recupero compiti: HTTP ${conn.responseCode}"))
                }
            } catch (e: Exception) {
                Result.failure(e)
            }
        }

    override suspend fun getReportCard(token: String): Result<ScrutinyReportCard> =
        withContext(Dispatchers.IO) {
            try {
                val url = URL("$baseUrl/scrutiny/my-report-card")
                val conn = (url.openConnection() as HttpURLConnection).apply {
                    requestMethod = "GET"
                    setRequestProperty("Authorization", "Bearer $token")
                    setRequestProperty("Accept", "application/json")
                    connectTimeout = 10000
                    readTimeout = 10000
                }

                if (conn.responseCode in 200..299) {
                    val response = BufferedReader(InputStreamReader(conn.inputStream)).use { it.readText() }
                    val obj = JSONObject(response)
                    val gradesMap = mutableMapOf<String, Double>()
                    val gradesJson = obj.optJSONObject("grades")
                    gradesJson?.keys()?.forEach { key ->
                        gradesMap[key] = gradesJson.optDouble(key, 0.0)
                    }

                    Result.success(
                        ScrutinyReportCard(
                            studentId = obj.optString("student_id", ""),
                            period = obj.optInt("period", 2),
                            grades = gradesMap,
                            conductGrade = obj.optInt("conduct_grade", 8),
                            finalDecision = obj.optString("final_decision", "Ammesso"),
                            credits = obj.optInt("credits", 0)
                        )
                    )
                } else {
                    Result.failure(Exception("Errore recupero pagella: HTTP ${conn.responseCode}"))
                }
            } catch (e: Exception) {
                Result.failure(e)
            }
        }
}

File: ios/student/Services/StudentAPIService.swift
import Foundation

public protocol StudentAPIServiceProtocol {
    var lastStudentName: String? { get }
    func login(email: String, password: String) async throws -> String
    func fetchGrades(token: String) async throws -> [GradeItemModel]
    func fetchHomework(token: String) async throws -> [HomeworkItemModel]
    func fetchAttendance(token: String) async throws -> [StudentAttendanceRecordModel]
}

public class HttpStudentAPIService: StudentAPIServiceProtocol {
    private let baseURL: URL
    public var lastStudentName: String?

    public init(baseURL: URL? = nil) {
        if let baseURL = baseURL {
            self.baseURL = baseURL
        } else if let url = URL(string: AppConfig.baseURL) {
            self.baseURL = url
        } else {
            self.baseURL = URL(string: "https://registro-backend-fdu2.onrender.com/api/v1")!
        }
    }

    public func login(email: String, password: String) async throws -> String {
        var cleanEmail = email.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        // Trasparenza per alias documentati (es. studentea_1 -> studente2a_1)
        if cleanEmail.hasPrefix("studentea_") {
            cleanEmail = cleanEmail.replacingOccurrences(of: "studentea_", with: "studente2a_")
        } else if cleanEmail.hasPrefix("studenteb_") {
            cleanEmail = cleanEmail.replacingOccurrences(of: "studenteb_", with: "studente2b_")
        }

        let url = baseURL.appendingPathComponent("auth/login")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body = ["email": cleanEmail, "password": password]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse, (200...299).contains(httpResponse.statusCode) else {
            if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
               let errMsg = json["error"] as? String ?? json["message"] as? String {
                let code = (response as? HTTPURLResponse)?.statusCode ?? 401
                throw NSError(domain: "StudentAPI", code: code, userInfo: [NSLocalizedDescriptionKey: "\(errMsg) (HTTP \(code))"])
            }
            let code = (response as? HTTPURLResponse)?.statusCode ?? 401
            throw NSError(domain: "StudentAPI", code: code, userInfo: [NSLocalizedDescriptionKey: "Credenziali non valide o errore server (HTTP \(code))"])
        }

        if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] {
            let userObj = json["user"] as? [String: Any]
            let role = userObj?["role"] as? String
            guard role == nil || role == "student" else {
                throw NSError(domain: "StudentAPI", code: 403, userInfo: [NSLocalizedDescriptionKey: "Accesso non consentito: questo account non appartiene a uno studente."])
            }
            if let firstName = userObj?["first_name"] as? String, let lastName = userObj?["last_name"] as? String {
                self.lastStudentName = "\(firstName) \(lastName)".trimmingCharacters(in: .whitespacesAndNewlines)
            } else if let name = userObj?["first_name"] as? String ?? userObj?["name"] as? String {
                self.lastStudentName = name
            }

            if let token = json["token"] as? String ?? json["access_token"] as? String {
                return token
            }
        }
        throw NSError(domain: "StudentAPI", code: 500, userInfo: [NSLocalizedDescriptionKey: "Token non presente nella risposta"])
    }

    private func fetchSubjectsMap(token: String) async -> [String: String] {
        let url = baseURL.appendingPathComponent("subjects")
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        guard let (data, response) = try? await URLSession.shared.data(for: request),
              let httpResponse = response as? HTTPURLResponse, (200...299).contains(httpResponse.statusCode),
              let jsonArray = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            return [:]
        }

        var map: [String: String] = [:]
        for obj in jsonArray {
            if let id = obj["id"] as? String, let name = obj["name"] as? String {
                map[id] = name
            }
        }
        return map
    }

    public func fetchGrades(token: String) async throws -> [GradeItemModel] {
        let subjectMap = await fetchSubjectsMap(token: token)

        let url = baseURL.appendingPathComponent("grades/my-grades")
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse, (200...299).contains(httpResponse.statusCode) else {
            throw NSError(domain: "StudentAPI", code: 401, userInfo: [NSLocalizedDescriptionKey: "Non autorizzato"])
        }

        var rawGradesList: [[String: Any]] = []

        if let jsonObject = try? JSONSerialization.jsonObject(with: data) {
            if let directArray = jsonObject as? [[String: Any]] {
                rawGradesList = directArray
            } else if let dict = jsonObject as? [String: Any] {
                if let semesters = dict["semesters"] as? [[String: Any]] {
                    for semester in semesters {
                        if let sGrades = semester["grades"] as? [[String: Any]] {
                            rawGradesList.append(contentsOf: sGrades)
                        }
                    }
                } else if let grades = dict["grades"] as? [[String: Any]] {
                    rawGradesList = grades
                }
            }
        }

        return rawGradesList.enumerated().map { index, dict in
            let rawSubjectId = dict["subject_id"] as? String ?? ""
            let subjectName = (dict["subject_name"] as? String) ?? (dict["subject"] as? String) ?? subjectMap[rawSubjectId] ?? "Materia"

            let gradeVal: Double
            if let d = dict["grade_value"] as? Double {
                gradeVal = d
            } else if let i = dict["grade_value"] as? Int {
                gradeVal = Double(i)
            } else if let d = dict["grade"] as? Double {
                gradeVal = d
            } else if let i = dict["grade"] as? Int {
                gradeVal = Double(i)
            } else {
                gradeVal = 0.0
            }

            let weightVal: Double
            if let w = dict["weight"] as? Double {
                weightVal = w
            } else if let i = dict["weight"] as? Int {
                weightVal = Double(i)
            } else {
                weightVal = 1.0
            }

            let rawType = (dict["evaluation_type"] as? String) ?? (dict["grade_type"] as? String) ?? (dict["type"] as? String) ?? "Orale"
            var rawDate = (dict["date"] as? String) ?? ""
            if rawDate.count >= 10 {
                rawDate = String(rawDate.prefix(10))
            }

            return GradeItemModel(
                id: "\(dict["id"] ?? index)",
                subject: subjectName,
                grade: gradeVal,
                weight: weightVal,
                type: rawType,
                date: rawDate
            )
        }
    }

    public func fetchHomework(token: String) async throws -> [HomeworkItemModel] {
        let url = baseURL.appendingPathComponent("agenda")
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse, (200...299).contains(httpResponse.statusCode) else {
            return []
        }

        guard let jsonArray = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            return []
        }

        return jsonArray.enumerated().map { index, dict in
            var dueDate = (dict["due_date"] as? String) ?? (dict["date"] as? String) ?? ""
            if dueDate.count >= 10 {
                dueDate = String(dueDate.prefix(10))
            }
            return HomeworkItemModel(
                id: "\(dict["id"] ?? index)",
                subject: (dict["subject_name"] as? String) ?? (dict["subject"] as? String) ?? (dict["title"] as? String) ?? "Compito",
                taskDescription: (dict["description"] as? String) ?? (dict["notes"] as? String) ?? "",
                dueDate: dueDate,
                isCompleted: (dict["is_completed"] as? Bool) ?? (dict["completed"] as? Bool) ?? false
            )
        }
    }

    public func fetchAttendance(token: String) async throws -> [StudentAttendanceRecordModel] {
        let url = baseURL.appendingPathComponent("attendance/my-attendance")
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse, (200...299).contains(httpResponse.statusCode) else {
            return []
        }

        guard let jsonArray = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            return []
        }

        return jsonArray.enumerated().map { index, dict in
            var rawDate = (dict["date"] as? String) ?? ""
            if rawDate.count >= 10 {
                rawDate = String(rawDate.prefix(10))
            }
            return StudentAttendanceRecordModel(
                id: "\(dict["id"] ?? index)",
                date: rawDate,
                type: (dict["type"] as? String) ?? (dict["status"] as? String) ?? "Assenza",
                isJustified: (dict["is_justified"] as? Bool) ?? (dict["justified"] as? Bool) ?? false
            )
        }
    }
}
